source/inference.md

```diff
@@ -283,45 +255,23 @@ expect our sample proportions from this population to vary for samples of size 4
 ```
 
 We again use the `sample` to take samples of size 40 from our
-population of Airbnb listings. But this time we use a for loop
-to take 20,000 samples of size 40.
+population of Airbnb listings. But this time we use a list comprehension
+to repeat an operation multiple time (as in the previous chapter).
+In this case we are taking 20,000 samples of size 40
+and to make it clear which rows in the data frame come
+which of the 20,000 samples,
+we also add a column called `replicate` with this information.
 
 ```{code-cell} ipython3
-:tags: [remove-cell]
-
-# We again use the `rep_sample_n` to take samples of size 40 from our
-# population of Airbnb listings. But this time we set the `reps` argument to 20,000 to specify
-# that we want to take 20,000 samples of size 40. \index{rep\_sample\_n!reps argument}
-# \index{rep\_sample\_n!size argument}
-```
-
-```{code-cell} ipython3
-np.random.seed(1)
-
-samples = []
-for rep in range(20000):
-    sample = airbnb.sample(40)
-    sample = sample.assign(replicate=rep)
-    samples.append(sample)
-samples = pd.concat([samples[i] for i in range(len(samples))])
-
+samples = pd.concat([airbnb.sample(40).assign(replicate=n) for n in range(20_000)])
```

Comment:
This is what I suggest we use instead of the loop. I considered making the replicate numbers start at 1 as in the R book, but decided it was unnecessary visual noise just to make it more like R.

Reply:
So the original is:
```Python
samples = []
for rep in range(20000):
    sample = airbnb.sample(40)
    sample = sample.assign(replicate=rep)
    samples.append(sample)
samples = pd.concat([samples[i] for i in range(len(samples))])
```

and the replacement is very good:

```Python
samples = pd.concat([airbnb.sample(40).assign(replicate=n) for n in range(20_000)])
```

but could we maybe split onto two lines as in:

```Python
dfs = [airbnb.sample(40).assign(replicate=n) for n in range(20_000)]
samples = pd.concat(dfs)
```

While I was thinking about this, I wondered if for loops are not allowed,
then maybe we can simplify the accidental complexity by using a function:

```Python
def sample(data, n):
    return data.sample(n)
samples_list = [sample(data, n).assign(replicate=rep) for rep in range(20000)]
samples = pd.concat(samples_list)
```

and while we're at it, maybe make the function do the whole infer-like chunk-of-dataframe generation:

```Python
def getsample(data, n, rep):
    sample = data.sample(n)
    samplerows = sample.assign(replicate=rep)
    return samplerows
samples = pd.concat(getsample(data, n, rep) for rep in range(20000))
```


To be clear, I think the current re-write is fine, just throwing ideas out there for possible simplifications.

Reply:
Oh yeah and Hello, I'm Ivan. I recently learned about he UBC data science courses and I think y'all are doing a great job. I'm particularly interested in the Python parts (bcs I'm only beginner in R). I was following this repo and thought I'd jump in to offer some suggestions. 

I'm dealing with similar issues around using Python to teach stats using the modern curriculum (simulation, permutations tests, and bootstrap estimation), but also not trying to teach all of Python (because students don't identify as "coders" so we don't want to alienate them from learning stats).

I'm starting to think learning Python is inevitable though.... It seems to me that it will take longer to teach learner X statistics directly, than to offer the same learner a [Python tutorial](https://nobsstats.com/tutorials/python_tutorial.html), after which they can learn [modern statistics](https://minireference.com/blog/wp-content/uploads/2021/03/TheIntroductoryStatisticsCourse__Cobb.pdf) based on computational methods based on Python+pandas+scipy.stats+viz(alt,sns,plotnine).

Reply:
Hello @ivanistheone, thank your for taking the time to chime in here. We don't teach functions in this course we can't use those suggestions unfortunately. Regarding breaking up into two steps, I think that is a good idea when we first introduce list comprehensions in the previous chapter, which I am going to review now and keep your suggestion in mind if there is a similar situation occurring there (I also added a sentence to clarify here). I wouldn't personally break it down in two steps on such as big comprehension to avoid storing the data twice, but I think it is good in a toy example for educational purposes.

Thanks again for your comments and kind words about the UBC data science courses!

Reply:
+1 -- thanks for your input @ivanistheone !